### ISSUE TITLE
android_386_cgo causes a linker error

### ISSUE THREAD

**Issue by ekuiter** [bazelbuild/rules_go#2371]
### What version of rules_go are you using?

v0.21.2 (latest)

### What version of gazelle are you using?

None

### What version of Bazel are you using?

2.1.0 (latest)

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

Vagrant ubuntu/xenial64 (Linux ubuntu-xenial 4.4.0-145-generic #171-Ubuntu SMP Tue Mar 26 12:43:40 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux)

### Any other potentially useful information about your toolchain?

See the [bug repository](https://github.com/ekuiter/bazel_go_android_issue) for reproduction with Vagrant.

### What did you do?

Clone [this repository](https://github.com/ekuiter/bazel_go_android_issue), run `vagrant up` and `vagrant ssh`, cd into `/vagrant/example` and run `./build.sh`.

This should compile a very basic Go example for Android, but does not work.

(Background: I want to compile some Go code (https://github.com/scionproto/scion) for Android. In the end, I will need to allow for some C code, so I have to use the "@io_bazel_rules_go//go/toolchain:android_*_cgo" family of platforms.)

### What did you expect to see?

Compilation success when running:

```
bazel build //example --crosstool_top=@androidndk//:default_crosstool --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --cpu=x86 --platforms=@io_bazel_rules_go//go/toolchain:android_386_cgo
```

### What did you see instead?

A linker error:

```
internal/cpu.processOptions: relocation target __x86.get_pc_thunk.dx not defined
internal/cpu.processOptions: relocation target __x86.get_pc_thunk.ax not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.ax not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.si not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.cx not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.bx not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.bp not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.dx not defined
internal/cpu.doinit: relocation target __x86.get_pc_thunk.di not defined
cmpbody: relocation target __x86.get_pc_thunk.cx not defined
memeqbody: relocation target __x86.get_pc_thunk.cx not defined
runtime.f32hash: relocation target __x86.get_pc_thunk.ax not defined
runtime.f64hash: relocation target __x86.get_pc_thunk.ax not defined
runtime.interhash: relocation target __x86.get_pc_thunk.dx not defined
runtime.interhash: relocation target __x86.get_pc_thunk.ax not defined
runtime.nilinterhash: relocation target __x86.get_pc_thunk.dx not defined
runtime.nilinterhash: relocation target __x86.get_pc_thunk.ax not defined
runtime.efaceeq: relocation target __x86.get_pc_thunk.dx not defined
runtime.efaceeq: relocation target __x86.get_pc_thunk.ax not defined
runtime.ifaceeq: relocation target __x86.get_pc_thunk.dx not defined
runtime.ifaceeq: relocation target __x86.get_pc_thunk.ax not defined
external/go_sdk/pkg/tool/linux_amd64/link: too many errors
```

For the other architectures supported by Android, see the README file in the linked repository.

I have tried several version combinations for Bazel, rules_go, Android NDK, and Go. Nothing seems to work. The only way I can manage to compile the minimal example (as well as the actual code I want to compile, https://github.com/scionproto/scion) for Android is like this:

```
bazel build //example --platforms=@io_bazel_rules_go//go/toolchain:linux_386
```

However, this way no C code can be included with the application at all. Is it really impossible to compile a Go application with C bindings for Android, or am I missing something obvious?

**Comment by jayconrod** [comment]
cc @steeve who I think had this working in the past.

**Comment by steeve** [comment]
Indeed, we are running Go + CGo on all Android platforms (minus mips of course). Let me take a look.

**Comment by steeve** [comment]
It seems it has to do with the linkmode.
Indeed, I can successfully build the example with `linkmode = "c-shared"` (as we are using it). So there is indeed something.

**Comment by steeve** [comment]
This must be an issue with the way https://github.com/golang/go/blob/9f3c2b6d16bd00f77d8c92801425809d4d767baa/src/cmd/go/internal/work/init.go#L94-L190 is implemented in rules_go.

**Comment by steeve** [comment]
Indeed: Android can only run PIE executables.
You just need to add `linkmode = "pie"` to you `go_binary` and it builds OK.

```
go_binary(
    name = "example",
    embed = [":go_default_library"],
    visibility = ["//visibility:public"],
    linkmode = "pie",
)
```

```
~/c/b/example (master)> bazel build //example --verbose_failures --crosstool_top=@androidndk//:default_crosstool --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --cpu=x86 --platforms=@io_bazel_rules_go//go/toolchain:android_386_cgo
INFO: Analyzed target //example:example (29 packages loaded, 9323 targets configured).
INFO: Found 1 target...
Target //example:example up-to-date:
  bazel-bin/example/android_386_stripped_pie/example
INFO: Elapsed time: 28.694s, Critical Path: 18.05s
INFO: 4 processes: 4 darwin-sandbox.
INFO: Build completed successfully, 7 total actions
```

Perhaps rules_go could do it automatically? Happens in https://github.com/bazelbuild/rules_go/blob/master/go/private/mode.bzl#L221

**Comment by jayconrod** [comment]
@steeve Thanks for investigating.

I'm working on a change that switches cross-compilation from using an aspect to using Bazel's new build settings and transitions. Mode validation should be part of that, or a follow-up soon after. Currently there are a lot of modes that can be specified, but aren't valid, and mode.bzl does not do a good job of reporting failures.

**Comment by steeve** [comment]
Ah it’s great you’re doing it! Faster analysis times and proper cross compilation from attributes! I’m super excited

**Comment by ekuiter** [comment]
Thank you, setting `linkmode = "pie"` did the trick for me, it even works with the larger codebase I was trying to compile. :)

**Comment by steeve** [comment]
Welcome to our wonderful world :D
